server/src/services/documentSymbol/SymbolVisitor.ts

```diff
@@ -0,0 +1,68 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { SymbolKind } from "vscode-languageserver-types";
+import _ from "lodash";
+import { RuleKind, TokenKind } from "@nomicfoundation/slang/kinds";
+import { RuleNode, TokenNode } from "@nomicfoundation/slang/cst";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { slangToVSCodeRange } from "../../parser/slangHelpers";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+
+export abstract class SymbolVisitor {
+  public abstract ruleKind: RuleKind;
+  public abstract symbolKind: SymbolKind;
+  public abstract nameTokenKind: TokenKind;
+
+  constructor(
+    public document: TextDocument,
+    public symbolBuilder: SymbolTreeBuilder
+  ) {}
+
+  public onRuleNode(node: RuleNode, cursor: Cursor): void {
+    const range = slangToVSCodeRange(this.document, cursor.textRange);
+
+    let symbolName = "-";
+    let selectionRange = slangToVSCodeRange(this.document, cursor.textRange);
+
+    // Find identifier
+    const childCursor = cursor.spawn();
+
+    do {
+      const identifierNode: TokenNode | null = childCursor.findTokenWithKind([
```

Comment:
suggestion, since this is doesn't have to be an identifier:
```suggestion
      const nameToken: TokenNode | null = childCursor.findTokenWithKind([
```